### PR TITLE
[None][refactor] decoding inputs, part 2

### PR DESCRIPTION
**Pull request by Funatiq**
## Description

- Remove `decoder_batch::Input` and related interfaces 
  - Modify `MakeDecodingBatchInputOutput` to accept `DecoderInputBuffers` directly, simplifying the interface.
  - Update `GptDecoderBatched` to work with the new `DecoderInputBuffers` structure.
- Adjust Python bindings to reflect changes in the C++ interface.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

## Summary by CodeRabbit

* **New Features**
  * Added Python bindings for decoder buffer management classes, enabling direct access to decoder input/output buffers and slot buffers from Python.
  * Exposed buffer structures for easier integration and manipulation in Python-based workflows.

* **Refactor**
  * Simplified and streamlined decoder input preparation by updating interfaces to modify buffers in place instead of returning new objects.
  * Unified and clarified buffer member usage and naming for logits and decoding steps across the codebase.
  * Removed deprecated or redundant classes and bindings, improving maintainability and consistency.
  * Updated decoder and runtime interfaces to use the new buffer structures, removing legacy input classes.

* **Bug Fixes**
  * Updated internal references to buffer members to ensure correct handling of logits and decoding steps.

* **Chores**
  * Integrated new buffer bindings into module initialization for both pybind11 and nanobind backends.
  * Added new source files for buffer bindings to build configurations.

### CODE REVIEW

**Comment by Funatiq** [comment]
/bot run --stage-list "A10-CPP-1, A30-CPP-1, A30-CPP-2, A30-CPP-3, H100_PCIe-CPP-1, H100_PCIe-CPP-2"

**Review by DomBrown — APPROVED** [review]
LGTM
